File: backend/app/models/category.py
import uuid
from sqlalchemy import Column, String, ForeignKey, Integer, DateTime
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base


class Category(Base):
    __tablename__ = "categories"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(36), ForeignKey("users.id"), nullable=False)
    name = Column(String(100), nullable=False)
    parent_id = Column(String(36), ForeignKey("categories.id"), nullable=True)
    icon = Column(String(50), nullable=True)
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    user = relationship("User", back_populates="categories")
    parent = relationship("Category", remote_side=[id], back_populates="children")
    children = relationship("Category", back_populates="parent", cascade="all, delete-orphan")
    reminders = relationship("Reminder", back_populates="category", cascade="all, delete-orphan")

File: frontend/lib/models/category.dart
class Category {
  final String id;
  final String name;
  final String? parentId;
  final String? icon;
  final int sortOrder;
  final List<Category> children;

  Category({required this.id, required this.name, this.parentId, this.icon, this.sortOrder = 0, this.children = const []});

  factory Category.fromJson(Map<String, dynamic> json) => Category(
    id: json['id'],
    name: json['name'],
    parentId: json['parent_id'],
    icon: json['icon'],
    sortOrder: json['sort_order'] ?? 0,
    children: (json['children'] as List<dynamic>? ?? []).map((e) => Category.fromJson(e)).toList(),
  );

  Map<String, dynamic> toJson() => {
    'name': name,
    'parent_id': parentId,
    'icon': icon,
    'sort_order': sortOrder,
  };
}
